From 110dc3bf3fbd731631ad69b8a83b94950be335f8 Mon Sep 17 00:00:00 2001
From: Zachary Zhou <zchrzhou@gmail.com>
Date: Sun, 27 Oct 2024 15:30:06 +0800
Subject: [PATCH 3/6] st: right-click pastes from the primary section

---
 src/config.def.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/config.def.h b/src/config.def.h
index 2cd740a..fca7c26 100644
--- a/src/config.def.h
+++ b/src/config.def.h
@@ -176,7 +176,7 @@ static uint forcemousemod = ShiftMask;
  */
 static MouseShortcut mshortcuts[] = {
 	/* mask                 button   function        argument       release */
-	{ XK_ANY_MOD,           Button2, selpaste,       {.i = 0},      1 },
+	{ XK_ANY_MOD,           Button3, selpaste,       {.i = 0},      1 },
 	{ ShiftMask,            Button4, ttysend,        {.s = "\033[5;2~"} },
 	{ XK_ANY_MOD,           Button4, ttysend,        {.s = "\031"} },
 	{ ShiftMask,            Button5, ttysend,        {.s = "\033[6;2~"} },
-- 
2.25.1

