From 75a5156e16f3c4e6e5d32a0e5c513a4ef5eea90f Mon Sep 17 00:00:00 2001
From: Zachary Zhou <zchrzhou@gmail.com>
Date: Sun, 27 Oct 2024 16:42:34 +0800
Subject: [PATCH 5/6] st: st-scrollback shift + mouse

---
 src/config.def.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/config.def.h b/src/config.def.h
index 33e72c8..e99e8c5 100644
--- a/src/config.def.h
+++ b/src/config.def.h
@@ -176,6 +176,8 @@ static uint forcemousemod = ShiftMask;
  */
 static MouseShortcut mshortcuts[] = {
 	/* mask                 button   function        argument       release */
+ 	{ ShiftMask,            Button4, kscrollup,      {.i = 1} },
+	{ ShiftMask,            Button5, kscrolldown,    {.i = 1} },
 	{ XK_ANY_MOD,           Button3, selpaste,       {.i = 0},      1 },
 	{ ShiftMask,            Button4, ttysend,        {.s = "\033[5;2~"} },
 	{ XK_ANY_MOD,           Button4, ttysend,        {.s = "\031"} },
-- 
2.25.1

